### PR TITLE
feat: Add command with `setattr`

### PR DESCRIPTION
**Pull request by KazuCocoa**
As https://github.com/appium/python-client/pull/608#issuecomment-856307740 and after I considered better way, I think we can add a new method dynamically with `setattr`. Then, the new command can call as `driver.new_command`. It should be better than call a method as string as the previous one.

I haven't updated docstring.

### CODE REVIEW

File: appium/webdriver/webdriver.py [hunk 1]
```diff
@@ -114,6 +114,43 @@ def _make_w3c_caps(caps: Dict) -> Dict[str, List[Dict[str, Any]]]:
...
 
+class ExtensionBase:
+    """
+    Used to define an extension command.
+
+    TODO:
+    add examples
+    """
+
+    def __init__(self, execute: Callable):
```
Comment (mykola-mokhnach) [inline appium/webdriver/webdriver.py:125]: Python also allows to make typing more precise with Generics, for example Callable[str, str] defines a method, which returns str and accepts one param of type str

File: appium/webdriver/webdriver.py [hunk 2]
```diff
@@ -114,6 +114,43 @@ def _make_w3c_caps(caps: Dict) -> Dict[str, List[Dict[str, Any]]]:
...
+        self._execute = execute
+
+    def execute(self, parameters: Any) -> Any:
```
Comment (mykola-mokhnach) [inline appium/webdriver/webdriver.py:128]: are parameters required?

File: appium/webdriver/webdriver.py [hunk 3]
```diff
@@ -114,6 +114,43 @@ def _make_w3c_caps(caps: Dict) -> Dict[str, List[Dict[str, Any]]]:
...
+        return self._execute(self.method_name(), parameters)
+
+    def method_name(self) -> str:
+        """
+        Expected to return a method name.
+        This name will be able to call as a driver method.
```
Comment (mykola-mokhnach) [inline appium/webdriver/webdriver.py:134]: will be available for calling

File: appium/webdriver/webdriver.py [hunk 4]
```diff
@@ -114,6 +114,43 @@ def _make_w3c_caps(caps: Dict) -> Dict[str, List[Dict[str, Any]]]:
...
+
+        Returns:
+            'str' The method name.
+        """
+        raise NotImplementedError()
+
+    def command_wrapper(self) -> Any:
```
Comment (mykola-mokhnach) [inline appium/webdriver/webdriver.py:141]: some examples would be nice to have for the next two methods

File: appium/webdriver/webdriver.py [hunk 5]
```diff
@@ -176,6 +214,16 @@ def __init__(
         By.IMAGE = MobileBy.IMAGE
         By.CUSTOM = MobileBy.CUSTOM
 
+        for extension in extensions:
+            instance = extension(self.execute)
+            if hasattr(WebDriver, instance.method_name()):
+                logger.warn(
+                    '{} is alreadt defined. The new one is overriding the old one.'.format(instance.method_name())
```
Comment (mykola-mokhnach) [inline appium/webdriver/webdriver.py:221]: already defined

Comment (mykola-mokhnach) [reply appium/webdriver/webdriver.py:221]: consider using f-strings instead

File: appium/webdriver/webdriver.py [hunk 6]
```diff
@@ -176,6 +214,16 @@ def __init__(
...
+                )
+            setattr(WebDriver, instance.method_name(), instance.command_wrapper)
```
Comment (mykola-mokhnach) [inline appium/webdriver/webdriver.py:223]: should we actually invoke `command_wrapper` here? I'm not sure about what this method is supposed to return

File: appium/webdriver/webdriver.py (hunk 5 above)
Comment (mykola-mokhnach) [reply appium/webdriver/webdriver.py:221]: overriding existing command might cause unexpected results. I would rather throw an exception here

File: test/unit/webdriver/webdriver_test.py [hunk 7]
```diff
@@ -254,70 +260,73 @@ def exceptionCallback(request, uri, headers):
 
     @httpretty.activate
     def test_add_command(self):
-        driver = ios_w3c_driver()
+        class CustomURLCommand(ExtensionBase):
+            def method_name(self):
+                return 'test_command'
+
+            def command_wrapper(self):
+                return self.execute({})['value']
```
Comment (mykola-mokhnach) [inline test/unit/webdriver/webdriver_test.py:268]: I'm not quite sure how are we supposed to pass custom arguments here?

File: appium/webdriver/webdriver.py (hunk 5 above)
Comment (KazuCocoa) [reply appium/webdriver/webdriver.py:221]: yea, I'm looking for instance wide inclusion like Ruby's `include`  to raise an error as previous add_command stuff.
`hasattr` seems class or loaded module wide

**Review by jlipps — COMMENTED** [review]
👀

File: Pipfile [hunk 8]
```diff
@@ -18,6 +18,7 @@ tox = "~=3.23"
 
 httpretty = "~=1.0"
 python-dateutil = "~=2.8"
+types-python-dateutil = "~=0.1"
```
Comment (KazuCocoa) [inline Pipfile:21]: Azure failure pointed to this necessity.

File: appium/webdriver/webdriver.py [hunk 9]
```diff
@@ -114,6 +114,113 @@ def _make_w3c_caps(caps: Dict) -> Dict[str, List[Dict[str, Any]]]:
...
+class ExtensionBase:
+    """
+    Used to define an extension command as driver's methods.
+
+    Example:
+        When you want to add `example_command` which calls a get request to
+        `session/$sessionId/path/to/your/custom/url`.
+
+        1. Defines an extension as a subclass of `ExtensionBase`
+            class CustomYourCommand(ExtensionBase):
```
Comment (jlipps) [inline appium/webdriver/webdriver.py:126]: `YourCustomCommand`

File: appium/webdriver/webdriver.py [hunk 10]
```diff
@@ -176,6 +284,24 @@ def __init__(
         By.IMAGE = MobileBy.IMAGE
         By.CUSTOM = MobileBy.CUSTOM
 
+        self._extensions = extensions
+        for extension in self._extensions:
+            instance = extension(self.execute)
```
Comment (jlipps) [inline appium/webdriver/webdriver.py:289]: we use `instance.method_name()` a lot. should we just make it a variable for readability?

```
name = instance.method_name()
```
or
```
method_name = instance.method_name()
```

**Review by jlipps — APPROVED** [review]
awesome!

File: appium/webdriver/webdriver.py [hunk 11]
```diff
@@ -114,6 +114,113 @@ def _make_w3c_caps(caps: Dict) -> Dict[str, List[Dict[str, Any]]]:
...
+
+            # Then, the driver calls a get request to `session/$sessionId/path/to/your/custom/$id/url`
+            # with replacing the `$id` with the given `element.id`
+            driver.custom_method_name(element.id)
+    """
+
+    def __init__(self, execute: Callable[[str, Dict], Dict[str, Any]]):
+        self._execute = execute
+
+    def execute(self, parameters: Any = {}) -> Any:
```
Comment (mykola-mokhnach) [inline appium/webdriver/webdriver.py:204]: Please do not set mutable objects as defaults: https://docs.quantifiedcode.com/python-anti-patterns/correctness/mutable_default_value_as_argument.html

Comment (mykola-mokhnach) [reply appium/webdriver/webdriver.py:204]: shall we use `*args, **kwargs` notation instead?

Comment (KazuCocoa) [reply appium/webdriver/webdriver.py]: Changed here as Dict explicitly to match with original execute: https://github.com/SeleniumHQ/selenium/blob/e46dba531896f67bdb93ba8d9e1d64f8b6b75b6f/py/selenium/webdriver/remote/webdriver.py#L380

**Pull request merged**